src/lower/mod.rs

```diff
@@ -445,6 +447,10 @@ impl LowerWhereClause<ir::DomainGoal> for WhereClause {
             WhereClause::TraitRefWellFormed { ref trait_ref } => {
                 ir::WellFormed::TraitRef(trait_ref.lower(env)?).cast()
             }
+            WhereClause::TyFromEnv { ref ty } => ir::FromEnv::Ty(ty.lower(env)?).cast(),
```

Comment:
I am torn on the question of what this should mean:

```
if (T: Trait) { .. }
```

I am thinking maybe we should make that equivalent to `if (FromEnv(T: Trait)) { .. }` and then have an explicit form `if (Implemented(T: Trait)) { .. }`. Basically `T: Trait` would parse to an AST form whose expansion depends on whether it appears in a "clause" (in which case it is `FromEnv`) or a Goal (in which case it is "Implemented").

Reply:
(That said, I'm not really sure when we would want to say `if (Implemented(T: Trait)) { ... }`.. but maybe it's useful when crafting unit tests or something.)

Reply:
Yeah not really sure either, but I think converting to `FromEnv(T: Trait)` in clauses is a good thing, and the explicit goal `Implemented` would replace the former `if_raw` statement in a much nicer way.